Merge Parser Part

ndaelman-hu opened the pull request:
Feel free to provide feedback in the form of reviews.

ndaelman-hu committed:
First notes

ndaelman-hu committed:
Flesh out `From Text to Hierarchies` section

ndaelman committed:
Flesh out intro

ndaelman committed:
- Add start section on `MappingParser`
- Restructure `From Text to Hierarchies`

ndaelman committed:
Clean up phrasing "From Hierarchies to Schema"

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+Likely one of NOMAD's most recognized and beloved features is _drag-and-drop parsing_.
+We understand that annotating data, keeping track of all steps involved, is most often a time-consuming, painful process.
+Still, good _research data management_ is central to good scientific reporting.
```
> Still, good _research data management_ is central to good scientific reporting.

Do we want to replace the first use of "good" with "effective"?

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+## Modular Design and Plugins
+
+Historically, our parsers have come always come together as a neat bundle.
```
> Historically, our parsers have come always come together as a neat bundle.

Historically, our parsers have ~~come~~ always come together as a neat bundle.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+Historically, our parsers have come always come together as a neat bundle.
+As interest in research data management has become more widespread and recognized across scientific disciplines, NOMAD has adapted along with those interests.
+Nowadays, you can find besides the [central NOMAD cloud service](), also standalone NOMAD Oasis deployments geared towards the needs of individual institutes or research groups.
```
> Nowadays, you can find besides the [central NOMAD cloud service](), also standalone NOMAD Oasis deployments geared towards the needs of individual institutes or research groups.

Nowadays, you can find, besides the [central NOMAD cloud service](), ~~also~~ standalone NOMAD Oasis deployments geared towards the needs of individual institutes or research groups.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+### Getting started
+
+Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.
```
> Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.

Getting hands-on with a plugin is ~~just~~ as easy as visiting our [GitHub template]() and _using one in your own codespace_.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+### Getting started
+
+Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.
+The template will appear almost empty at the start, but following the instructions `README.md`  and during the `cruft` setup will allow you to tune the project to your needs.
```
> The template will appear almost empty at the start, but following the instructions `README.md`  and during the `cruft` setup will allow you to tune the project to your needs.

The template will appear ~~almost empty~~ bare-bones at the start, ~~but~~ following the instructions in `README.md`  and ~~during~~ the `cruft` setup will allow you to tune the project to your needs.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.
+The template will appear almost empty at the start, but following the instructions `README.md`  and during the `cruft` setup will allow you to tune the project to your needs.
+Just a couple of notes on the latter:
```
> Just a couple of notes on the latter:

Just a couple of notes on ~~the latter~~ `cruft`:

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+The template will appear almost empty at the start, but following the instructions `README.md`  and during the `cruft` setup will allow you to tune the project to your needs.
+Just a couple of notes on the latter:
+
+- NOMAD uses the Apache 2.0 license. Unless your organization prohibits this, please select the same license for maximal legal compatibility.
```
> NOMAD uses the Apache 2.0 license. Unless your organization prohibits this, please select the same license for maximal legal compatibility.

NOMAD uses the Apache 2.0 license. ~~Unless your organization prohibits this, please~~ Please select the same license for maximal legal compatibility unless restricted by your organization.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+- NOMAD uses the Apache 2.0 license. Unless your organization prohibits this, please select the same license for maximal legal compatibility.
+- Parsers require both a `parsers` and a `schema_packages` folder. The reason for the second folder will soon become apparent.
+- We use generically named files like `parsers/myparser.py` as blueprints by copying them to new, more clearly named files. Each file format, its own Python file.
```
> Each file format, its own Python file.

This sentence is unclear to me. Each file format has its own Python file? Each file format needs its own Python file?

ndaelman-hu replied:
Each parser for a file format has its own Python file, yes.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+### Managing Entry Points
+
+The plugin setup actually follows a more common Python standard from the `importlib.metadata`, namely [entry-points](https://setuptools.pypa.io/en/latest/userguide/entry_point.html).
```
> The plugin setup actually follows a more common Python standard from the `importlib.metadata`, namely [entry-points](https://setuptools.pypa.io/en/latest/userguide/entry_point.html).

The plugin setup ~~actually~~ follows a more common Python standard from the `importlib.metadata`, ~~namely~~ [entry-points](https://setuptools.pypa.io/en/latest/userguide/entry_point.html).

JFRudzinski replied:
The plugin setup follows the common entry-points Python standard from importlib.metadata.

ndaelman-hu replied:
Slightly modified your suggestion:

"The plugin setup follows a more common Python standard of [entry-points](https://setuptools.pypa.io/en/latest/userguide/entry_point.html) provided by `importlib.metadata`."

ndaelman-hu replied:
Just saw your comment now, Joseph

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+### Managing Entry Points
+
+The plugin setup actually follows a more common Python standard from the `importlib.metadata`, namely [entry-points](https://setuptools.pypa.io/en/latest/userguide/entry_point.html).
+It enables `pip` installing select functionalities from a module, instead of the whole.
```
> It enables `pip` installing select functionalities from a module, instead of the whole.

Select functionalities from a module can be installed via `pip`, installing the entire module is not required.

JFRudzinski replied:
I like Bernadette's suggestion but I think it should still start with "It enables" since you are referring to the entry-points setup:
"""
It enables pip to install select functionalities from a module, without installation of the entire package.
"""
not sure if package is the right word, so you can replace that

ndaelman-hu replied:
Ended up going a different, clearer phrasing here:

"It works in tandem with `pip` install to allow for a more elegant and controlled way of exposing and loading (specific functionalities in) modules."

ndaelman-hu replied:
Same here, think these comments dropped after me resolving them.
I will stick with my updated wording, since this is correcter.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+It enables `pip` installing select functionalities from a module, instead of the whole.
+Entry points also provide the module developer how these functionalities ought to be exposed to the environment, e.g. their name and own configuration.
+
+Conceptually, there are five key players that you should keep track off:
```
> Conceptually, there are five key players that you should keep track off:

Conceptually, there are five key players that you ~~should~~ need to keep track of ~~f~~:

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+- the **module setup file** _exposes_ the entry point (and its group) under the format of `<module_name>.<functionality_folder_name>:<entry_point_name>`. This is the name by which you should call this entry point. In NOMAD we use the `pyproject.toml` setup file under the module's root folder.
+- the **NOMAD configuration file** called `nomad.yaml` allows further control over which entry points to _include_ or _exclude_, as well as their _configuration parameters_.
+
+Given a typical folder lay-out for a parser project, the players are located in
```
TODO?

ndaelman-hu replied:
What do you mean?

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+## From Hierarchies to Schema
+
+As parsing involves the mapping between two well-defined formats, one could expect it to be just that trivial.
+In practice, however, parser developers have to manage discrepancies in semantics, shape, data type or units.
```
Do we want to briefly introduce terms such as parser, schema, semantics, etc., at the beginning of the tutorial? Depending on the audience, this might be helpful.

ndaelman-hu replied:
Good point! Given that I go 3rd, @JFRudzinski can maybe explain _semantics_ and @JosePizarro3 _schema_?
I will of course define _parser_, which I believe I already do here in the text.

JFRudzinski replied:
schema will certainly be clear by this point. For semantics, not sure if there is a clear spot for me to explain, but we can come back to this.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+4. reshape and mangle the data to match the target/NOMAD `Quantity`s' specification. This may include computing derived properties not present in the original source files.
+5. build up a Python `archive` object using the classes provided by the target/NOMAD schema.
+
+In the past, we encountered cases where the delineation between these responsibilities blurred, resulting in widely different parser designs.
```
> In the past, we encountered cases where the delineation between these responsibilities blurred, resulting in widely different parser designs.

In the past, we encountered cases where the delineation between these responsibilities were blurred, resulting in widely different parser designs.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+In the past, we encountered cases where the delineation between these responsibilities blurred, resulting in widely different parser designs.
+Especially when dealing with the larger, more feature-rich parsers, these designs could become quite complex.
+
+During refactoring, we therefore addressed these responsibilities on a more individual basis.
```
> During refactoring, we therefore addressed these responsibilities on a more individual basis.

During a recent ~~refactoring~~ redesign/restructuring/overhaul ... ?

JFRudzinski replied:
+1

ndaelman-hu replied:
Is _refactoring_ too technical?

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+During refactoring, we therefore addressed these responsibilities on a more individual basis.
+This lead to a clear defined _interfaces_ and powerful _tools_ for the developer.
+These are further complimented by _best practices_ and _design protocols_.
```
> This lead to a clear defined _interfaces_ and powerful _tools_ for the developer.

We now provide clearly defined _interfaces_ and powerful _tools_ for the developer.

ndaelman committed:
- Write out "Parsing Hierarchical Tree Formats"
- Link to other parts
- TODO: resolve comments in "Parsing Hierarchical Tree Formats"

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+More on the exact "how" later.
+
+Lastly, we will cover tools for point 5 (building the `archive`) and 2 (targeting data fields) in the next subsections.
+Point 5, in particularl, was a developing nuisance, as it required the reconstruction of the already defined target/NOMAD schema.
```
> Point 5, in particularl, was a developing nuisance, as it required the reconstruction of the already defined target/NOMAD schema.

Point 5, in ~~particularl~~ particular, ~~was a developing nuisance, as it~~ previously required the reconstruction of the already defined target/NOMAD schema.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+Lastly, we will cover tools for point 5 (building the `archive`) and 2 (targeting data fields) in the next subsections.
+Point 5, in particularl, was a developing nuisance, as it required the reconstruction of the already defined target/NOMAD schema.
+That comes across as doing double work: running over the schema once to define it, and a second time to initialize it.
```
> That comes across as doing double work: running over the schema once to define it, and a second time to initialize it.

Do we need this sentence?

ndaelman-hu replied:
I'm a bit torn:

I don't want to unnecessarily throw any shade at ourselves.
However, I do want to emphasize how impactful and pragmatic this addition is. The reader might think: "well, it's just a small instantiation step", but in practice is creates a lot of repition and new headaches (source: believe me).

I guess the compromise would be to reformulate this sentence better. Any suggestions?

JFRudzinski replied:
I would personally remove both this sentence and "Point 5, in particular, ...". The attendees do not need to know the whole back story IMO. If they think it's easier than it is, we are doing our job well :)

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+That comes across as doing double work: running over the schema once to define it, and a second time to initialize it.
+The main hurdle to automation here was the variable structure of the source data:
+a well-defined hierarchical format does not ensure that individual file structures remains constant.
+Some leaf nodes or branches may simply not be present.
```
> Some leaf nodes or branches may simply not be present.

Do people know what we mean by _leaf nodes_ and _branches_ ?

ndaelman-hu replied:
It's common teminology when dealing with trees. If they don't, then the entire tree anology falls apart anyhow. We could add a visual that subtly specifies the concepts.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+With all of these points covered, developing a parser now comes down to just implementing point 3 (matching source with target).
+This is the stage where data actually becomes semantically enriched and standardized.
+This is a profoundly human task (barring any further leaps in LLM capabilities), where domain expertise is most needed.
```
> This is a profoundly human task (barring any further leaps in LLM capabilities), where domain expertise is most needed.

At present, this is a profoundly human task ~~(barring any further leaps in LLM capabilities)~~ where domain expertise is most needed.

ndaelman-hu replied:
And here I was inserting messages on our doom by singularity, but you called me out :D
But yeah, I shouldn't change register like that.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+With all of these points covered, developing a parser now comes down to just implementing point 3 (matching source with target).
+This is the stage where data actually becomes semantically enriched and standardized.
+This is a profoundly human task (barring any further leaps in LLM capabilities), where domain expertise is most needed.
+It is thus this scientific curation task that is now most emphasized in parser development, rather than the coding aspects.
```
> It is thus this scientific curation task that is now most emphasized in parser development, rather than the coding aspects.

Therefore, this scientific curation task is now most emphasized in parser development.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+The leaf nodes, meanwhile, can be of either two types:
+
+- `Map`: the default mapping from as source leaf/branch to a target leaf/branch. Both of these are expressed as Paths according to the [jmespath specifications]().
+Hint: paths may be relative wrt a parent. The format is  to (a) start the path with the connector, i.e. dot (.); and (b) denote the parent's Path separately.
```
> Hint: paths may be relative wrt a parent.

Hint: paths may be relative ~~wrt~~ _w.r.t._/_to_ a parent.

ndaelman-hu replied:
Completely rephrased this part.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+## From Text to Hierarchies
+
+Contrary to XML or HDF5, text files do not come with a machine-readable structure.
```
> Contrary to XML or HDF5, text files do not come with a machine-readable structure.

Contrary to XML or HDF5, plain text files do not ~~come with~~ have a machine-readable structure.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+
+Contrary to XML or HDF5, text files do not come with a machine-readable structure.
+Therefore, step 2 (targeting data fields) requires an intermediate step to fall back on _key matching_.
+The source format is converted to a temporary tree format stored just in RAM by matching lines via regular expressions.
```
> The source format is converted to a temporary tree format stored just in RAM by matching lines via regular expressions.

The source format is converted to a temporary tree format stored ~~just~~ in ~~RAM~~ memory by matching lines via regular expressions.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+You can add a match group to a pattern via `capture(quantity)`.
+
+For blocks, the matching group should lie between the header and footer.
+Most times, devs prefer to approach the matching group content in a block as generic, non-descript text to focus on the stand-out characteristics of the block itself.
```
> Most times, devs prefer to approach the matching group content in a block as generic,

Most times, ~~devs~~ developers prefer to approach the matching group content in a block as generic,

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+For blocks, the matching group should lie between the header and footer.
+Most times, devs prefer to approach the matching group content in a block as generic, non-descript text to focus on the stand-out characteristics of the block itself.
+You can use the `FHIAimsOutParser.re_non_greedy` flag here.
+Be careful: it should be superceded by a final regex pattern, else the matching continues up to the end-of-file.
```
> Be careful: it should be superceded by a final regex pattern, else the matching continues up to the end-of-file.

Be careful: it should be ~~superceded~~ superseded by a final regex pattern ~~, else~~ ; otherwise, the matching continues ~~up~~ to the end of the file.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+The actual parsing of the data then happens at a deeper layer.
+It is therefore not uncommon to encounter block parsing of up to three layers deep.
+
+Sometimes, it is easier to map the data by their column index.
```
> Sometimes, it is easier to map the data by their column index.

It can be easier to map the data by their column index depending on their shape.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+Focus first on identifying the line as whole, which always end on `FHIAimsOutParser.re_eol`.
+Ensure that you toggle the `xxxx.Quantity.repeats` option to `True` or any non-zero number, so you match all lines.
+The actual parsing of the data then happens at a deeper layer.
+It is therefore not uncommon to encounter block parsing of up to three layers deep.
```
> up to three layers deep

Is this a set maximum depth, or should we keep it general, e.g. _several layers deep_ ?

ndaelman-hu replied:
Uhm, the matching will continue going deeper until it reaches a final `Quantity` level. Here I was presenting a single step in that process, but one can keep on nesting of course. Will think on how to clarify this.

Bernadette-Mohr commented on docs/parser_plugins.md:
```diff
@@ -1 +1,176 @@
+This is just a convenient short-hand.
+Accessing the data works just as with any other nested structure.
+
+Following this logic through, you will end up with a continuation of a vertically connected tree.
```
> Following this logic through, you will end up with a continuation of a vertically connected tree.

Following this logic ~~through~~ , you will end up with a continuation of a vertically connected tree.

Bernadette-Mohr reviewed (commented):
All my comments are suggestions and everything is up for debate!

ndaelman committed:
Cover different source - target path lengths

ndaelman committed:
Restructure source - ttarget path lengths

ndaelman-hu commented:
> All my comments are suggestions and everything is up for debate!

Thx! Most of these seem like spelling or grammar corrections, so I'll readily apply them.

Any feedback on concepts, or structure? As agreed upon, the tutorial (and presentation in particular) should focus on introducing concepts. It  should also be accessible to complete novices.

ndaelman-hu commented:
I did include some technical / how-to aspects when writing these docs. Whether I cover any during the presentation depends on the time at hand.

Still, in the docs, they are intermingled with the concepts, which are our main focus. I just address points as they naturally pop up. What do you say if I **use admonishments to hide away the techinical paragraphs in folded boxes**?

JosePizarro3 commented:
@ndaelman I am already checking, and have some points we must discuss:
1. Please, stop overusing emphasis, bolding, etc. It came to a point that almost each paragraph has some word or even phrase with it.
2. The introduction is far too much. I'll suggest to go more to the point here. And discuss with @JFRudzinski if moving certain parts to Part 1, but rewritten in a different style.

I will focus now on the practical parts.

Bernadette-Mohr commented:
> > All my comments are suggestions and everything is up for debate!
> 
> Thx! Most of these seem like spelling or grammar corrections, so I'll readily apply them.
> 
> Any feedback on concepts, or structure? As agreed upon, the tutorial (and presentation in particular) should focus on introducing concepts. It should also be accessible to complete novices.

Personally, I would try to clearly highlight what is relevant for parsers and how we implement it now. I would try to really put the spotlight on the essence of what we want to communicate.

Bernadette-Mohr commented:
> I did include some technical / how-to aspects when writing these docs. Whether I cover any during the presentation depends on the time at hand.
> 
> Still, in the docs, they are intermingled with the concepts, which are our main focus. I just address points as they naturally pop up. What do you say if I **use admonishments to hide away the techinical paragraphs in folded boxes**?

I think this is a good idea. Limit the top level of the presentation/document to what is really necessary to get the point across. Additional details can be provided depending on time/audience/questions etc. Similar to a presentation at an interdisciplinary conference.

JFRudzinski commented:
I am working on the review now...

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
\ No newline at end of file
+# Creating Parser Plugins from Scratch
+
+Likely one of NOMAD's most recognized and beloved features is _drag-and-drop parsing_.
```
I suggest a reorg and trimming of this first paragraph:
"""
One of NOMAD's most recognized features is drag-and-drop parsing. The NOMAD parsers, which automate the conversion of raw simulation files into the standardized NOMAD format, significantly offload the burden of data annotation from researchers, reducing their data management responsibilities while also improving the accessibility of their data.
"""

ndaelman-hu replied:
Yeah, I like this repoloshing better. Thx for improving my draft. Will keep it in the same place.

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+Parsers _automate the conversion_ of a file format common to a specific piece of software or a community into the NOMAD format.
+The schema for computational techniques was introduced in the previous section.
+
+The key advantages of the NOMAD schema are summed up in **FAIR**mat's core values:
```
Let's move this to the written part of the introduction. I will also link my slides, but was planning to write some overview for future reference. I think this would fit best there. You could just cut and paste it over there and I can organize and extend later.

ndaelman-hu replied:
Will move it `intro.md` on my branch, okay?

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+## Modular Design and Plugins
+
+Historically, our parsers have come always come together as a neat bundle.
```
I think this should also be moved to the introduction. I also have some edits, but I think it's easier if you just move it there and then I can edit on a future branch

ndaelman-hu replied:
Have moved all of "Modular Design and Plugins".

ndaelman-hu replied:
It is somewhat parser-specific, though. Maybe you want to open up the paragraph to all plugins.

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+### Getting started
+
+Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.
```
I would replace this first sentence with something like:
"""
To create your own parser plugin, visit our GitHub template and use the “Use this template” button to create your own project. Follow the [How to get started with plugins](https://nomad-lab.eu/prod/v1/staging/docs/howto/plugins/plugins.html) documentation for detailed setup instructions.
"""

ndaelman-hu replied:
Less sales-pitchy, more action-oriented. Kay

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+## Modular Design and Plugins
+
+Historically, our parsers have come always come together as a neat bundle.
```
No need to add context about "history of NOMAD".  You can essentially start this docu page with:

"
In this section, we will show you how to write a parser plugin for your simulation data. In the context of simulations, a NOMAD parser is a set of Python modules that have the responsibility of reading the output and input files of a program or software, and mapping data into a specific data schema. As you saw in Part II - NOMAD-Simulations, we write this data into the schema defined in the `nomad-simulations` package. The plugin aspect of a parser comes from the modularity of NOMAD, in which each individual software should have an associated plugin with it (e.g., the `nomad-parser-vasp` is a central parser plugin for NOMAD).
"

ndaelman-hu replied:
This has been addressed in Joseph's review too.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+Conversely, it also lowers the threshold for whomever wants to contribute to this open-source community: you can focus on just a single schema or parser.
+The possibilities are as wide as you can make them, we just facilitate the common basis.
+
+In the rest of this page, we are going to take you on a birds-eye overview of how a parser developer approaches these plugins.
```
Delete this paragraph.

JosePizarro3 replied:
And add a second paragraph commenting on "extending the schema in the parser plugin", with the help of @Bernadette-Mohr

ndaelman-hu replied:
> And add a second paragraph commenting on "extending the schema in the parser plugin"

I thought this has its own page. I actually refer to it.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+### Getting started
+
+Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.
```
Please, be more direct. 
"
In order to generate a parser plugin with the option of extending the schema, go to the [nomad-plugin-template] and click on the top right, "Use this template". You can follow the instructions on the README.md of the newly generated plugin, but importantly, **remember to add the option of plugin and schema_package**.
"

ndaelman-hu replied:
Also addressed in the previous review.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+### Getting started
+
+Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.
+The template will appear almost empty at the start, but following the instructions `README.md`  and during the `cruft` setup will allow you to tune the project to your needs.
```
`cruft` is a bit annoying to use, maybe some people has to pip install in a venv. How did you do it @ndaelman ?

ndaelman-hu replied:
`cruft` isn't for installing, but rather populating the template

JosePizarro3 replied:
I mean, how did you create the repo, generated via cruft, etc?

ndaelman-hu replied:
Yes. I don't see the confusion: these are the steps described in the pulgin-template.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+Getting hands-on with a plugin is just as easy visiting our [GitHub template]() and _using one in your own codespace_.
+The template will appear almost empty at the start, but following the instructions `README.md`  and during the `cruft` setup will allow you to tune the project to your needs.
+Just a couple of notes on the latter:
```
You can delete this points. Just add the explanation on splitting the files into specific submodules "parser.py", "parser_subfile2.py" etc.

ndaelman-hu replied:
> You can delete this points.

Given the previous comment, I shouldn't apparently.

> Just add the explanation on splitting the files into specific submodules

That's the last bulletpoint in the list.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+- Parsers require both a `parsers` and a `schema_packages` folder. The reason for the second folder will soon become apparent.
+- We use generically named files like `parsers/myparser.py` as blueprints by copying them to new, more clearly named files. Each file format, its own Python file.
+
+### Managing Entry Points
```
Why would you explain the whole functionality of EntryPoints here? Just explain that this is a structure done to avoid problems with circular imports when defining default plugins in NOMAD. Also, the last point on the `nomad.yaml` goes beyond the scope of this, but it should be rephrased better, like pointing to the general NOMAD docu on "how to link a plugin with your NOMAD"

ndaelman-hu replied:
> Also, the last point on the nomad.yaml goes beyond the scope

Meanwhile, @JFRudzinski is asking me to elaborate this point. I'll link the docs, though.

> Why would you explain the whole functionality of EntryPoints here?

Idt that it was one to avoid circular imports, but nice. Regardless, this is now the system by which any plugin is exposed (by definition). NOMAD requires it to pick up on them. In the end, that's all the audience needs to know.
(Yes, you can add certain dependencies as optional to the `pyproject.toml`, but to my understanding this does not apply to plugins.)

Having recently gone over the entry points, our docs were quite confusing (will give some feedback to Area D later). I definitely think that there's merit in explaning the structure, both textually and visually.
With that said, I'll hide the more in-depth discussion via an admonishment.

ndaelman-hu replied:
Migrated most of the entry point discussion to a section Extra at the bottom.
I retain 1 short paragraph in "Getting Started".

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+- Parsers require both a `parsers` and a `schema_packages` folder. The reason for the second folder will soon become apparent.
+- We use generically named files like `parsers/myparser.py` as blueprints by copying them to new, more clearly named files. Each file format, its own Python file.
+
+### Managing Entry Points
```
This section has some nice overview tips which are great supplements to the current docs, but I think this should be moved either down towards the end of this section.

The entry-points concepts have to do with linking your parser to some local NOMAD instance/Oasis, but we are not there yet. For me the steps are:
1. create your own project with the template
2. link your project to some nomad-lab and nomad-simulations packages via the toml (note the link to nomad-lab is for use of the classes and utilities not to integrate your parser yet)
3. write and test your parser (as standalone)
4.  Finally, you can now integrate your parser into, e.g., your Oasis

You can of course start with the integration explanation, but I think it will be much more difficult for newcomers and may overwhelm them

ndaelman-hu replied:
> I think this should be moved either down towards the end of this section.

How about the text box suggestion? I prefer keeping the thematic info in a single place, but hiding the tech part.

> For me the steps are

Step 1: check; step 2: will expand; step 3: is the majority of this part.
I disagree with the "standalone" parser: it always needs to be plugged into NOMAD or the Oasis, else it cannot read the mainfiles. Actually, I would recommend that people first connect it, since they may get confused on the origin of an error once they add their own code on top.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+Given a typical folder lay-out for a parser project, the players are located in
+...
+
+note on file matching
```
Still, the most important thing is left out. You explained the whole thing of what an EntryPoint is (totally unnecessary), but you didn't say specifically how to match a file, what is a mainfile, etc.

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+## From Hierarchies to Schema
+
+As parsing involves the mapping between two well-defined formats, one could expect it to be just that trivial.
```
As parsing involves the mapping between two well-defined formats, one could expect it to be trivial.

JosePizarro3 replied:
Again, too much blahblah. Go to the point!

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+In practice, however, parser developers have to manage discrepancies in semantics, shape, data type or units.
+This has lead to five distinct categories of responsibility for the developer to manage:
+
+1. select the relevant files and read them into Python.
```
I like this list. Please, this is what this Part should be about: these points must be the sub-sections of this docu page.

ndaelman-hu replied:
Thx. These are definitely the main concepts to show in the presentation.
If they, or their solutions become more fleshed out, I'll consider making them subsections in their own right.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+4. reshape and mangle the data to match the target/NOMAD `Quantity`s' specification. This may include computing derived properties not present in the original source files.
+5. build up a Python `archive` object using the classes provided by the target/NOMAD schema.
+
+In the past, we encountered cases where the delineation between these responsibilities blurred, resulting in widely different parser designs.
```
Delete this

ndaelman-hu replied:
I'd rather rephrase: just want to emphasize how misunderstanding the 5 points can lead to uncomfortable parser design. Feel free to make a suggestion.

JosePizarro3 replied:
If you want, you can add a smaller section at the end with "# Hints for developing parsers", but here it sounds weird. Too much talk, and not any real action in this document.

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+As parsing involves the mapping between two well-defined formats, one could expect it to be just that trivial.
+In practice, however, parser developers have to manage discrepancies in semantics, shape, data type or units.
+This has lead to five distinct categories of responsibility for the developer to manage:
```
I like this. Shortening these points and placing them into a small diagram for easier intake could be nice (not a priority)

ndaelman-hu replied:
I REALLY like my bullet points :D
Bernadette is working on this, so I'll hold off.

ndaelman-hu commented:
> I am working on the review now...

Will hold off on pushing then.

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+This lead to a clear defined _interfaces_ and powerful _tools_ for the developer.
+These are further complimented by _best practices_ and _design protocols_.
+
+More specifically, point 1 (file selection) is now handled by the `MatchingParser` class, from which the parser class inherits. It coordinates with the NOMAD base to scan the uploaded folder tree and select the relevant files.
```
I'm afraid that it is a bit difficult overwhelming to take in all these details without even having started any example. I think we need to have a bit more of a (extended) how-to framework, then the MatchingParser class will naturally come up as you set up the mock parser. Is this possible?

ndaelman-hu replied:
The problems tackled and their solutions can be separated, yes.
I'll prep some code snippet + introduce a subsection for the solutiuons. Bernadette's figure may also help here.

ndaelman-hu replied:
Turned the solutions into a list of their own. Will now add a short example showing solutions 1, 2, and 5 working together.

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+
+Let us cover the remaining tools, so we can move on to the real science.
+
+### Parsing Hierarchical Tree Formats
```
At this point the reader still has no idea how anything in the plugin works in practice. You mentioned the template but we haven't worked with it at all yet. We need some section/s before this which explain the basics before you start talking about hierarchical data trees, like how to link nomad-simulations and import the schema and populate a quantity in the most basic way.

ndaelman-hu replied:
I was thinking of adding the folder structure once the template is set up. Would that address your concerns?

ndaelman-hu replied:
I think that the new additions now show enough of the folder structure and parser code to provide a clear view.
Will resolve this thread, but feel free to open it if you disagree.

JFRudzinski reviewed (approved):
Thanks for all your work Nathan. I think you have some nice content here, this lengthier version just needs some refinement, see suggested edits for a start.

More generally, I am starting to think that we basically need a section before this section to just talk about the basic setup of the plugin, like linking and importing parts of the schema and the most basic population of a quantity. 

One option is that I simply take the H5MD parser and strip it down to almost nothing to demonstrate these things. But let's see first what you say to these comments and also Chema's comments.

JosePizarro3 commented on docs/parser_plugins.md:
```diff
@@ -1 +1,210 @@
+As interest in research data management has become more widespread and recognized across scientific disciplines, NOMAD has adapted along with those interests.
+Nowadays, you can find besides the [central NOMAD cloud service](), also standalone NOMAD Oasis deployments geared towards the needs of individual institutes or research groups.
+
+As such, NOMAD is shifting to a _plugin model_, where each deployment administrator can customize their setup.
```
Delete this paragraph

ndaelman-hu replied:
Has been moved to `intro.md`

JosePizarro3 reviewed (changes requested):
Hey @ndaelman, 

Ok, some comments on the overall structure which require a lot of changes. But I have the feeling that writing each of the paragraphs in your current version was much harder than the idea I propose:
- Remove everything except for "Parsing Hierarchical Tree Formats". This section on annotation is good.
- Start directly with a very naive paragraph similar to what I suggested.
- After that, explain in 1-2 paragraphs + 1-2 pictures how to create a plugin from the nomad-plugin-template. Remark here on the saying "yes" to add parser and schema_packages. And no need for so much details.
- Then, enter in a section "# Structure of the parser plugin" and explain the MatchingParser, the options, and how to do it for a example for which we need to add some real files into the docu (either if it VASP, FHIaims, or H5MD, this is whatever we can decide together). Note this is point 1 of your list in "From Hierarchies to Schema".
- Continue building from points 2 to 5 (maybe you need to readjust a bit the ordering): "# Extracting data from raw files", "# Populating the `nomad-simulations` schema", "# Appending `Simulation` to the archive of an entry"...
- If you feel you need to add an explanation of this annotations system, I suggest you add it before or after. This is the background / tool which users will use, but its concepts and details are more left for the audience who needs to get into the fine details. Still, I think some explanation is good (that's why I suggested to leave the current section now), but the most important thing are the concepts of mapping from files to the schema. 

And finally, be more assertive; users won't care about how historically things were, or other phrases uses along the docu, but they want actual "Do this" and "Do that" with a single paragraph explaining stuff in a complete and concise way.

ndaelman committed:
Apply Bernadette's feedback

ndaelman committed:
Apply Joseph's feedback

ndaelman committed:
Move intro to `intro.md`

ndaelman committed:
- Streamline "From Hierarchies to Schema"
- Move entry-points section down
- TODO: alter title under "From Hierarchies to Schema"

ndaelman committed:
- Show how to connect `MatchingParser`, `XMLParser`, and `MetainfoParser`
- Rename some subsection titles

ndaelman committed:
Polish "Getting Started"

ndaelman committed:
Streamline "From Hierarchies to Schema

ndaelman committed:
Rephrase listing under "Extra: aligning Source and Target Segments"

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+They should not be edited or exposed directly.
+
+The `README.md` should remain a copy of the template, in case anyone wants to fork the project.
+Instead, place plugin descriptions in the entry points or GitHub repository metadata.
```
I think "instead" doesn't quite make sense:
"""
You can place any basic plugin descriptions in the ..."

JFRudzinski commented on docs/parser_plugins.md:
````diff
@@ -1 +1,264 @@
+Lastly, NOMAD uses the Apache 2.0 license (option 4 in the `cruft` setup).
+Please select the same license for maximal legal compatibility.
+
+```bash
````
I would do one of the following options with the example, either 

1. just move it above directly under "By the end, your repository should look like the example below:" and then continue that paragraph underneath

or 2. move it to the end of the subsection separated from the text

JFRudzinski commented on docs/parser_plugins.md:
````diff
@@ -1 +1,264 @@
+├── nomad.yaml                   "(NOMAD configuration file)"
+```
+
+This examples also highlights the files containing _entry point_ information between parentheses.
````
"The above example"

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+This examples also highlights the files containing _entry point_ information between parentheses.
+They are explained in greater detail in section [[Extra: Managing Entry Points]].
+For now, it suffices to understand them as the official mechanism for `pip` installing modules.
+The dependence runs bottom up, i.e. the `module setup file` refers to `entry point` that refers to `target functionality`.
```
"...which then refers to target functionality"

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+The dependence runs bottom up, i.e. the `module setup file` refers to `entry point` that refers to `target functionality`.
+
+The final entry point piece (`NOMAD configuration file`) resides in `nomad.yaml` and refers back to the `module setup file`.
+Normally, you don't have to touch anything there for NOMAD to pick up on your parser.
```
"In most cases, you don't ..."?

"...to recognize your parser"?

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+The final entry point piece (`NOMAD configuration file`) resides in `nomad.yaml` and refers back to the `module setup file`.
+Normally, you don't have to touch anything there for NOMAD to pick up on your parser.
+It does allows you to control the loading of plugins and their options.
+Closing the "dependence" cycle, `myparser.py` has to read in these options.
```
"dependency"?

This last sentence is a little unclear to me. What exactly do you mean by "has to read"

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+Here, they are ordered to match the parser's execution:
+
+1. **file selection** - navigate the upload's folder structure and select the relevant files.
+2. **source extraction** - read the files into Python. This step may already include some level of data field filtering.
```
data-field

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+
+1. **file selection** - navigate the upload's folder structure and select the relevant files.
+2. **source extraction** - read the files into Python. This step may already include some level of data field filtering.
+3. **source to target** - map the data of interest with their counterparts in the target/NOMAD schema. This is where the bulk of the filtering happens.
```
"...to its appropriate location in the target (NOMAD) schema." ?

you should use "map to" in either case. I think "counterparts" would be the right word if you were saying that the mapping was between 2 data structures, but in this case you are mapping a piece of data (which lives in some structure) into a new structure, right? (This is a subtle language thing, not sure if I am correct here)

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+1. **file selection** - navigate the upload's folder structure and select the relevant files.
+2. **source extraction** - read the files into Python. This step may already include some level of data field filtering.
+3. **source to target** - map the data of interest with their counterparts in the target/NOMAD schema. This is where the bulk of the filtering happens.
+4. **data mangling** - manipulate the data to match the target/NOMAD `Quantity`s' specification, e.g. dimensionality, shape. This may include computing derived properties not present in the original source files.
```
I think we can drop "target" now and just say NOMAD

"... NOMAD's `Quantity` specification..."

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+2. **source extraction** - read the files into Python. This step may already include some level of data field filtering.
+3. **source to target** - map the data of interest with their counterparts in the target/NOMAD schema. This is where the bulk of the filtering happens.
+4. **data mangling** - manipulate the data to match the target/NOMAD `Quantity`s' specification, e.g. dimensionality, shape. This may include computing derived properties not present in the original source files.
+5. **archive construction** - build up a Python `EntryArchive` object using the classes provided by the target/NOMAD schema. NOMAD will automatically write it commit it to the database as an `archive.json`
```
drop "target"

"...automatically write this object as an `archive.json` and commit it to the database."?

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+4. **data mangling** - manipulate the data to match the target/NOMAD `Quantity`s' specification, e.g. dimensionality, shape. This may include computing derived properties not present in the original source files.
+5. **archive construction** - build up a Python `EntryArchive` object using the classes provided by the target/NOMAD schema. NOMAD will automatically write it commit it to the database as an `archive.json`
+
+Blurring these responsibilities leads to a wild-growth in parser design and added complexity, especially in larger, more feature-rich parsers.
```
"...leads to a wild growth in parser complexity"

not sure what you meant with "parser design",  "design challenges"? you could add "complexity and [something]", but also I think only complexity is fine

"...larger, feature-rich parsers"

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+NOMAD therefore offers powerful _tools_ and documentation on _best practices_ to help the parser developer manage each distinct responsibility.
+The exact solutions are, in the same order:
+
+1. `MatchingParser` - This class passes the file contents on to the parser class, which extends it via inheritance. Since it interfaces with the NOMAD base directly, it will read most of the settings automatically in from there. Options may be tweaked via the entry-steps mechanism. 
```
"...it will read in most of the settings automatically from there."

JFRudzinski replied:
entry-steps?

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+The exact solutions are, in the same order:
+
+1. `MatchingParser` - This class passes the file contents on to the parser class, which extends it via inheritance. Since it interfaces with the NOMAD base directly, it will read most of the settings automatically in from there. Options may be tweaked via the entry-steps mechanism. 
+2. `XMLParser` and co. / `TextParser` - There are several _reader classes_ for loading common source formats into Python data types. Examples include the `XMLParser` and `HDF5Parser`. We will demonstrate `XMLParser` in [[Parsing Hierarchical Tree Formats]]. Plain text files, meanwhile, involve an additional _matching step_ via the `TextParser`. More on this in [[From Text to Hierarchies]].
```
"`XMLParser`, `TextParser`, etc." OR "File parser classes, e.g., `XMLParser` or `TextParser`" OR "File reader classes (e.g., ....)" OR "Reader classes (e.g., ...)"

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+
+1. `MatchingParser` - This class passes the file contents on to the parser class, which extends it via inheritance. Since it interfaces with the NOMAD base directly, it will read most of the settings automatically in from there. Options may be tweaked via the entry-steps mechanism. 
+2. `XMLParser` and co. / `TextParser` - There are several _reader classes_ for loading common source formats into Python data types. Examples include the `XMLParser` and `HDF5Parser`. We will demonstrate `XMLParser` in [[Parsing Hierarchical Tree Formats]]. Plain text files, meanwhile, involve an additional _matching step_ via the `TextParser`. More on this in [[From Text to Hierarchies]].
+3. `MappingAnnotationModel` - This is arguably the most involved part for the parser developer, as this is where the external data gets further semantically enriched and standardized. It requires doamin expertise to understand the relationship between the data fields in the source file and the [NOMAD-Simulations schema](nomad_simulations.md). If step 2 went well, this step only involves _annotati ...
```
typo "domain"

JFRudzinski replied:
"...via the `MappingAnnotationModel` class"

JFRudzinski replied:
I think there is a little bit missing here, including possible a disclaimer about this class and what the alternative was/is:

"...NOMAD-Simulations schema. This can be achieved through a "manual" population of each piece of data into the NOMAD `EntryArchive` object. However, this can be tedious and often leads to complex parser structures that are difficult to extend. For this reason, the `MappingAnnotationModel` class is currently being developed to automatically perform the necessary data mapping through simple _annotations_. More on this..."

Just a first suggestion, you likely need to adjust it

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+1. `MatchingParser` - This class passes the file contents on to the parser class, which extends it via inheritance. Since it interfaces with the NOMAD base directly, it will read most of the settings automatically in from there. Options may be tweaked via the entry-steps mechanism. 
+2. `XMLParser` and co. / `TextParser` - There are several _reader classes_ for loading common source formats into Python data types. Examples include the `XMLParser` and `HDF5Parser`. We will demonstrate `XMLParser` in [[Parsing Hierarchical Tree Formats]]. Plain text files, meanwhile, involve an additional _matching step_ via the `TextParser`. More on this in [[From Text to Hierarchies]].
+3. `MappingAnnotationModel` - This is arguably the most involved part for the parser developer, as this is where the external data gets further semantically enriched and standardized. It requires doamin expertise to understand the relationship between the data fields in the source file and the [NOMAD-Simulations schema](nomad_simulations.md). If step 2 went well, this step only involves _annotati ...
+4. [NOMAD-Simulations schema](nomad_simulations.md) / `MappingAnnotationModel` - The `MSection`s and `utils.py` in the schema provide _normalizers_ and _helper functions_ to alleviate most of the data mangling. For small amendments*, use `MappingAnnotationModel.operator`. For larger ones, consider extending the schema as covered in [Extending NOMAD-Simulations](schema_plugins.md).
```
For terminology like MSections and such we should reference the main docs if possible

JFRudzinski replied:
ammendments*?

JFRudzinski commented on docs/parser_plugins.md:
```diff
@@ -1 +1,264 @@
+2. `XMLParser` and co. / `TextParser` - There are several _reader classes_ for loading common source formats into Python data types. Examples include the `XMLParser` and `HDF5Parser`. We will demonstrate `XMLParser` in [[Parsing Hierarchical Tree Formats]]. Plain text files, meanwhile, involve an additional _matching step_ via the `TextParser`. More on this in [[From Text to Hierarchies]].
+3. `MappingAnnotationModel` - This is arguably the most involved part for the parser developer, as this is where the external data gets further semantically enriched and standardized. It requires doamin expertise to understand the relationship between the data fields in the source file and the [NOMAD-Simulations schema](nomad_simulations.md). If step 2 went well, this step only involves _annotati ...
+4. [NOMAD-Simulations schema](nomad_simulations.md) / `MappingAnnotationModel` - The `MSection`s and `utils.py` in the schema provide _normalizers_ and _helper functions_ to alleviate most of the data mangling. For small amendments*, use `MappingAnnotationModel.operator`. For larger ones, consider extending the schema as covered in [Extending NOMAD-Simulations](schema_plugins.md).
+5. `MetainfoParser` - this _converter_ bridges the annotated schema from step 3 with the reader classes in step 2. `MetainfoParser.data_object` contains the final `ArchiveSection` that is stored under `archive.data`.
```
"converter class"?

JFRudzinski reviewed (approved):
Here's a review up to "Assembling a Parser Class", I will continue the rest in a 2nd review.

Great work!! :1st_place_medal: This first part reads very nicely now. I just have a series of smaller typo and grammatical things, and a few suggestions for clarity

ndaelman-hu commented:
Closing this PR in favor of #4
